src/lightning/fabric/utilities/cloud_io.py

```diff
@@ -70,3 +75,31 @@ def _atomic_save(checkpoint: Dict[str, Any], filepath: Union[str, Path]) -> None
     torch.save(checkpoint, bytesbuffer)
     with fsspec.open(filepath, "wb") as f:
         f.write(bytesbuffer.getvalue())
+
+
+def is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False) -> bool:
+    """Check if a path is directory-like.
+
+    This function determines if a given path is considered directory-like, taking into account the behavior
+    specific to the `gcsfs` library. For other filesystems, it behaves similarly to the standard `fs.isdir` method.
+
+    Args:
+        fs: The filesystem to check the path against.
+        path: The path or URL to be checked.
+        strict: A flag specific to GCS. If set to ``False``, any non-existing path is considered
+            as a valid directory-like path. In such cases, the directory (and any non-existing parent directories)
+            will be created on the fly. Defaults to False.
+    """
+    # GCS fsspec's is inconsistent with other file systems because GCS does not have real directories,
+    # see https://gcsfs.readthedocs.io/en/latest/api.html?highlight=makedirs#gcsfs.core.GCSFileSystem.mkdir
```

Comment:
S3 also doesn't have real directories 🤔. But there is no special mention about the behavior in their docs: https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem.isdir

Reply:
I have no experience with S3 but it would not surprise me that every object storage works like that. I can update with a S3 check as well

Reply:
I also added Azure Blob storage to the check.